### PR TITLE
Poseidon Hash function Updates and refactoring

### PR DESCRIPTION
**Pull request by nemothenoone**

### CODE REVIEW

File: include/nil/crypto3/hash/detail/poseidon/poseidon_mds_matrix.hpp [hunk 1]
```diff
@@ -80,6 +82,15 @@ namespace nil {
 #endif
                     inline mds_matrix_type generate_mds_matrix() {
                         mds_matrix_type new_mds_matrix;
+                        if (policy_type::full_rounds == 55) { // TODO: kimchi specialization
```
Comment (SK0M0R0H) [inline include/nil/crypto3/hash/detail/poseidon/poseidon_mds_matrix.hpp:85]: This is a draft switch between neptun/kimchi versions. We definitely need something else.

File: test/poseidon.cpp [hunk 2]
```diff
@@ -92,102 +97,221 @@ void hash_test_data(const TestSet &test_set) {
...
+    field_type::value_type expected_result = 0x0CA2C3342C2959D7CD94B5C9D4DC55900F5F60B345F714827C8B907752D5A209_cppui256;
+    sponge_construction_type pallas_sponge;
+    
+    pallas_sponge.absorb(input);
+    field_type::value_type res = pallas_sponge.squeeze();
+
+    BOOST_CHECK(res == expected_result);
+}
+
+/*BOOST_DATA_TEST_CASE(poseidon_strengthen_1, string_data("internal", "strengthen", "1"), data_set) {
```
Comment (SK0M0R0H) [inline test/poseidon.cpp:205]: Because of how we set the kimchi constants, we cannot run the usual Poseidon.

**Review by SK0M0R0H — CHANGES_REQUESTED** [review]

File: include/nil/crypto3/hash/detail/poseidon/poseidon_mds_matrix.hpp (hunk 1 above)
Comment (martun) [reply include/nil/crypto3/hash/detail/poseidon/poseidon_mds_matrix.hpp]: I have refactored most of the code, please review again.

File: test/poseidon.cpp (hunk 2 above)
Comment (martun) [reply test/poseidon.cpp]: I have refactored most of the code, please review again.

**Review by nkaskov — APPROVED** [review]

**Review by SK0M0R0H — APPROVED** [review]

**Pull request merged**